File: Backend/src/utils/templates/python/binary_tree.py
from typing import *
import sys, json, collections, traceback

# ###USERCODE###

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def __build_tree(vals):
    if not vals or vals[0] is None:
        return None
    root = TreeNode(vals[0])
    queue = collections.deque([root])
    i = 1
    while queue and i < len(vals):
        node = queue.popleft()
        if i < len(vals) and vals[i] is not None:
            node.left = TreeNode(vals[i])
            queue.append(node.left)
        i += 1
        if i < len(vals) and vals[i] is not None:
            node.right = TreeNode(vals[i])
            queue.append(node.right)
        i += 1
    return root

def __serialize_tree(root):
    if not root:
        return "null"
    result = []
    queue = collections.deque([root])
    while queue:
        node = queue.popleft()
        if node:
            result.append(str(node.val))
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append("null")
    # Strip trailing nulls
    while result and result[-1] == "null":
        result.pop()
    return "[" + ",".join(result) + "]"

def __format(result):
    if result is None:
        return "null"
    if isinstance(result, bool):
        return "true" if result else "false"
    if isinstance(result, TreeNode):
        return __serialize_tree(result)
    if isinstance(result, list):
        return json.dumps(result)
    if isinstance(result, float):
        return f"{result:.5f}"
    return str(result)

if __name__ == '__main__':
    raw = sys.stdin.read().strip()
    lines = [l.strip() for l in raw.splitlines() if l.strip()]
    args = []
    for line in lines:
        try:
            val = json.loads(line)
        except Exception:
            val = line
        # Convert list args that represent trees
        args.append(val)

    # Build TreeNode args where the value is a list
    final_args = []
    for arg in args:
        if isinstance(arg, list) and (len(arg) == 0 or arg[0] is None or isinstance(arg[0], (int, float, type(None)))):
            final_args.append(__build_tree(arg))
        else:
            final_args.append(arg)

    try:
        obj = Solution()
        # Entry point is replaced at runtime by code-runner.ts
        result = getattr(obj, '###ENTRYPOINT###')(*final_args)
        print(__format(result))
    except Exception:
        sys.stderr.write(traceback.format_exc())
        sys.exit(1)

File: Backend/src/utils/templates/javascript/binary_tree.js
// ###USERCODE###

class TreeNode {
    constructor(val = 0, left = null, right = null) {
        this.val = val; this.left = left; this.right = right;
    }
}

function __buildTree(vals) {
    if (!vals || vals.length === 0 || vals[0] === null) return null;
    const root = new TreeNode(vals[0]);
    const queue = [root];
    let i = 1;
    while (queue.length && i < vals.length) {
        const node = queue.shift();
        if (i < vals.length && vals[i] !== null) {
            node.left = new TreeNode(vals[i]);
            queue.push(node.left);
        }
        i++;
        if (i < vals.length && vals[i] !== null) {
            node.right = new TreeNode(vals[i]);
            queue.push(node.right);
        }
        i++;
    }
    return root;
}

function __serializeTree(root) {
    if (!root) return 'null';
    const result = [];
    const queue = [root];
    while (queue.length) {
        const node = queue.shift();
        if (node) {
            result.push(String(node.val));
            queue.push(node.left);
            queue.push(node.right);
        } else {
            result.push('null');
        }
    }
    while (result.length && result[result.length - 1] === 'null') result.pop();
    return '[' + result.join(',') + ']';
}

function __format(result) {
    if (result === null || result === undefined) return 'null';
    if (typeof result === 'boolean') return result ? 'true' : 'false';
    if (result instanceof TreeNode) return __serializeTree(result);
    if (Array.isArray(result)) return JSON.stringify(result);
    return String(result);
}

(function __main() {
    const lines = require('fs').readFileSync(0, 'utf-8').trim().split('\n').filter(Boolean);
    const args = lines.map(line => {
        try {
            const val = JSON.parse(line);
            // Convert arrays to TreeNode
            if (Array.isArray(val)) return __buildTree(val);
            return val;
        } catch (e) { return line; }
    });
    try {
        const obj = new Solution();
        const result = obj['###ENTRYPOINT###'](...args);
        process.stdout.write(__format(result) + '\n');
    } catch (e) {
        process.stderr.write((e.stack || String(e)) + '\n');
        process.exit(1);
    }
})();
